torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -2910,6 +2911,168 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
+  if (!lhs_has_bound || !rhs_has_bound) {
+    GRAPH_DEBUG(
+        "(SimplifierUnderContext) Final: ",
+        std::to_string(simplified_cmp_select_expr));
+    return simplified_cmp_select_expr;
+  }
+
+  analysis::BoundCompareResult cmp_res = analysis::BoundCompareResult::kEQ;
+  auto bound_solved = analysis::compareBound(
+      {lhs_bound.first, lhs_bound.second},
+      {rhs_bound.first, rhs_bound.second},
+      &cmp_res);
+  if (!bound_solved) {
+    GRAPH_DEBUG(
+        "(SimplifierUnderContext) Final: ",
+        std::to_string(simplified_cmp_select_expr));
+    return simplified_cmp_select_expr;
+  }
+
+  // Get the oppositive comparison operation
+  //    not(gt) = le, not(le) = gt, not(lt) = ge, not(ge) = lt
+  //    not(eq) = ne, not(ne) = eq
+  auto _not = [](const CompareSelectOperation& cmp_op) {
+    switch (cmp_op) {
+      case CompareSelectOperation::kGT:
+        return CompareSelectOperation::kLE;
+      case CompareSelectOperation::kGE:
+        return CompareSelectOperation::kLT;
+      case CompareSelectOperation::kLT:
+        return CompareSelectOperation::kGE;
+      case CompareSelectOperation::kLE:
+        return CompareSelectOperation::kGT;
+      case CompareSelectOperation::kNE:
+        return CompareSelectOperation::kEQ;
+      default:
+        TORCH_INTERNAL_ASSERT(cmp_op == CompareSelectOperation::kEQ)
+        return CompareSelectOperation::kNE;
+    }
+  };
+
+  // Check whether the compare result satisfies the given comparison operation.
+  auto _imply = [](const analysis::BoundCompareResult& cmp_res,
+                   const CompareSelectOperation& cmp_op) {
+    // a > b
+    auto gt_true = (cmp_res == analysis::BoundCompareResult::kGT);
+    // a >= b
+    auto ge_true =
+        (cmp_res == analysis::BoundCompareResult::kGE ||
+         cmp_res == analysis::BoundCompareResult::kGT ||
+         cmp_res == analysis::BoundCompareResult::kEQ);
+    // a < b
+    auto lt_true = (cmp_res == analysis::BoundCompareResult::kLT);
+    // a <= b
+    auto le_true =
+        (cmp_res == analysis::BoundCompareResult::kLE ||
+         cmp_res == analysis::BoundCompareResult::kLT ||
+         cmp_res == analysis::BoundCompareResult::kEQ);
+    // a == b
+    auto eq_true = (cmp_res == analysis::BoundCompareResult::kEQ);
+    // a != b
+    auto ne_true =
+        (cmp_res == analysis::BoundCompareResult::kLT ||
+         cmp_res == analysis::BoundCompareResult::kGT ||
+         cmp_res == analysis::BoundCompareResult::kNE);
+
+    switch (cmp_op) {
+      case CompareSelectOperation::kGT:
+        return gt_true;
+      case CompareSelectOperation::kGE:
+        return ge_true;
+      case CompareSelectOperation::kLT:
+        return lt_true;
+      case CompareSelectOperation::kLE:
+        return le_true;
+      case CompareSelectOperation::kNE:
+        return ne_true;
+      default:
+        TORCH_INTERNAL_ASSERT(cmp_op == CompareSelectOperation::kEQ)
+        return eq_true;
+    }
+  };
```

Comment:
This part is about to compare two bounds with regard to a cmp_op. I feel that it makes more sense to put this logic in the analysis (bounds_overlap.cpp) instead of the simplifier. In the analysis, you can encapsulate this into a function that takes bounds a, bounds b and a cmp_op as inputs, and returns a value that indicates 1) if the cmp_op always holds true for a and b; or 2) if the cmp_op always holds false for a and b; or 3) not determined. 
```
enum CmpOpEvalResult { true, false, notDetermined }
CmpOpEvalResult TORCH_API compareBound(
    const Bound& a,
    const Bound& b,
    CompareSelectOperation& cmp_op);
```  
Feel free to use better names if you decide to do so :) In this way, the logic in simplifier for "CompareSelectPtr" will be much easier, and the analysis can be reused in other places.

Reply:
I enabled `Bound` to support comparison operators including `<, >, <=, >=, ==, !=`. The comparison is conservative. `True` means all the elements satisfy the given comparison operator. But the `False` does not mean the opposite comparison operator is satisfied. It could be but not always.